[receiver/azuremonitor] Add support for azureauthextension (#39658)
open-telemetry/opentelemetry-collector-contrib: receiver/azuremonitorreceiver/README.md

- `subscription_ids`: list of subscriptions on which the resource's metrics are collected
- or `discover_subscriptions`: (default = `false`) If set to true, will collect metrics from all subscriptions in the tenant.

> The following settings are optional:azcore.TokenCredential

Reviewer: copy paste left over

> - `auth.authenticator`: Specifies the component ID to use to authenticate requests to azure monitor. It must implement interface `azcore.TokenCredential`.

Reviewer: ```suggestion
- `auth.authenticator`: Specifies the component ID to use to authenticate requests to Azure API. Use [azureauth extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/azureauthextension).
```

The following settings are optional:

- `auth.authenticator`: Specifies the component ID to use to authenticate requests to Azure API. Use [azureauth extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/azureauthextension).

> - (Deprecated, use `auth` instead) `credentials` (default = service_principal): Specifies the used authentication method. Supported values are `service_principal`, `workload_identity`, `managed_identity`, `default_credentials`.

Reviewer 2: ```suggestion
- `credentials` (Deprecated [v0.126.0]: use `auth` instead)(default = service_principal): Specifies the used authentication method. Supported values are `service_principal`, `workload_identity`, `managed_identity`, `default_credentials`.
```

IMO, it's easier to read when the setting's name comes first.

The following settings are optional:

> - `token_provider` or `auth` (default = service_principal): Specifies the used authentication method:

Reviewer: I think it can be clearer if you keep token_provider and auth separated bullet points. You can have a little duplication here no problem as I think I will remove auth later on.
Can you also put a deprecated flag on the auth one?

Author: Yes, I have added it in the README, and also added a warning in the code:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/356c20ca906cf6fc35f713334a5d32d23ae35264/receiver/azuremonitorreceiver/scraper.go#L189

- `client_id`

> Authentication using `azureauthextension` requires:

Reviewer: requires ``auth.authenticator`` ? I would even not add it as it sounds obvious.

    auth: "default_credentials"
```

[Using `azureauthextension`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/azureauthextension#azure-authenticator-extension) for authentication:

```yaml
receivers:
  azuremonitor:
    subscription_ids: ["${subscription_id}"]

>     token_provider: azureauth

Reviewer: I'm not sure that token_provider is the right term knowing that it does not generate token behind the scene. If you put that in others component though, I'm ok to keep same for consistency.

Author: So far it is not use in any other components. Do you have a field name that you prefer?

Author: I have changed to `credentials`. Is it better?

Reviewer: I'm discovering a bit, but it looks like there is a recurring pattern.
```yaml
auth:
  authenticator: <name of the extension>
```
![image](https://github.com/user-attachments/assets/261dcd28-42ac-4fe1-9e96-f09e225e1be7)

There is a common structure for this in the collector https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configauth/configauth.go, it's used for all authentications. 

That said this structure is dedicated to create http/grpc clients so what I would propose would be to create something like this: 

- In the azure auth extension, copy this structure, with the same logic, and with a getter to obtain the creds.
```go
// Authentication defines the auth settings for the azure receiver.
type Authentication struct {
	// AuthenticatorID specifies the name of the azure extension to use in order to authenticate the incoming data point.
	AuthenticatorID component.ID `mapstructure:"authenticator,omitempty"`
}

// GetAzureTokenCredentials attempts to select the credential of the appropriate azure extension from the list of extensions,
// based on the requested extension name. If an authenticator is not found, an error is returned.
func (a Authentication) GetAzureTokenCredentials(_ context.Context, extensions map[component.ID]component.Component) (*azcore.TokenCredential, error) {
	if ext, found := extensions[a.AuthenticatorID]; found {
		if cred, ok := ext.(azcore.TokenCredential); ok {
			return &cred, nil
		}
		return nil, errNotServer
	}

	return nil, fmt.Errorf("failed to resolve authenticator %q: %w", a.AuthenticatorID, errAuthenticatorNotFound)
}
```

- In the azure monitor receiver, add a field in config with that type, with a ``azureauth`` name (to avoid breaking change and it tells well what it is)

```yaml
azureauth:
  authenticator: azureauth/prd
```

WDYT?

Author: The `azureauth` extension has the support for `authenticator` if the receiver requires client/server config (see [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/58d4ce9f27658578f202c966d0d19f3b4e73a58e/extension/azureauthextension/extension.go#L32-L33)).

For this receiver, we actually need to use `azcore.TokenCredential`, which then implements [this function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/58d4ce9f27658578f202c966d0d19f3b4e73a58e/extension/azureauthextension/extension.go#L131).

So the easiest way would be to just replace the name `credentials` with `authenticator` in the receiver config ([here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2635e87cb0a2627d0d903899412b53beb23b6134/receiver/azuremonitorreceiver/config.go#L260)). But I don't know if `authenticator` would be the best name, as we are not really using it as a client/server.

Author: I can see that. To have

```
auth:
   authenticator: azureauth/prd
```

We need to first remove the current `auth`, otherwise it won't be possible.

Reviewer: I can make a breaking change before, to rename auth. Would it be preferable that way for you ? I'm anyway planning in the end to remove it

Author: Renaming could work, thanks!

Reviewer: I'm on it

Reviewer: Here it is https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39738

Reviewer: They already merged it <3
